overlord/state/copy_test.go

```diff
@@ -0,0 +1,122 @@
+
+	dstStateFile := filepath.Join(c.MkDir(), "dst-state.json")
+	err := ioutil.WriteFile(dstStateFile, nil, 0644)
+	c.Assert(err, IsNil)
+
+	err = state.CopyState(srcStateFile, dstStateFile, []string{"some-data"})
+	c.Assert(err, ErrorMatches, `cannot copy state: "/.*/dst-state.json" already exists`)
+}
+func (ss *stateSuite) TestCopyStateNoDataEntriesToCopy(c *C) {
+	srcStateFile := filepath.Join(c.MkDir(), "src-state.json")
+	dstStateFile := filepath.Join(c.MkDir(), "dst-state.json")
+
+	err := state.CopyState(srcStateFile, dstStateFile, nil)
+	c.Assert(err, ErrorMatches, `cannot copy state: must provide at least one data entry to copy`)
+}
+
+var srcStateContent = []byte(`
+{
+    "data": {
+        "api-download-tokens-secret": "123",
+        "api-download-tokens-secret-time": "2020-02-21T10:32:37.916147296Z",
+        "auth": {
+            "last-id": 1,
+            "users": [
+                {
+                    "id": 1,
+                    "email": "some@user.com",
+                    "macaroon": "1234",
+                    "store-macaroon": "5678",
+                    "store-discharges": [
+                        "9012345"
+                    ]
+                }
+            ],
+            "device": {
+                "brand": "generic",
+                "model": "generic-classic",
+                "serial": "xxxxx-yyyyy-",
+                "key-id": "xxxxxx",
+                "session-macaroon": "xxxx"
+            },
+            "macaroon-key": "xxxx="
+        },
+        "config": {
+        }
+    }
+}
+`)
+
+const stateSuffix = `,"changes":{},"tasks":{},"last-change-id":0,"last-task-id":0,"last-lane-id":0}`
+
+func (ss *stateSuite) TestCopyStateIntegration(c *C) {
+	// create a mock srcState
+	srcStateFile := filepath.Join(c.MkDir(), "src-state.json")
+	err := ioutil.WriteFile(srcStateFile, srcStateContent, 0644)
+	c.Assert(err, IsNil)
+
+	// copy
+	dstStateFile := filepath.Join(c.MkDir(), "dst-state.json")
+	err = state.CopyState(srcStateFile, dstStateFile, []string{"auth.users", "no-existing-does-not-error", "auth.last-id"})
+	c.Assert(err, IsNil)
+
+	// and check that the right bits got copied
+	dstContent, err := ioutil.ReadFile(dstStateFile)
+	c.Assert(err, IsNil)
+	c.Check(string(dstContent), Equals, `{"data":{"auth":{"last-id":1,"users":[{"id":1,"email":"some@user.com","macaroon":"1234","store-macaroon":"5678","store-discharges":["9012345"]}]}}`+stateSuffix)
+}
+
+var srcStateContent1 = []byte(`{
+    "data": {
+        "A": {"B": [{"C": 1}, {"D": 2}]},
+        "E": {"F": 2, "G": 3},
+        "H": 4,
+        "I": null
+    }
+}`)
+
+func (ss *stateSuite) TestCopyState(c *C) {
+	srcStateFile := filepath.Join(c.MkDir(), "src-state.json")
+	err := ioutil.WriteFile(srcStateFile, srcStateContent1, 0644)
+	c.Assert(err, IsNil)
+
+	dstStateFile := filepath.Join(c.MkDir(), "dst-state.json")
+	err = state.CopyState(srcStateFile, dstStateFile, []string{"A.B", "no-existing-does-not-error", "E.F", "E", "I"})
```

Comment:
Could you also add a check for non-existing path under an existing entry, e.g. "E.F.non-existing" and "E.non-existing"?